### PR TITLE
(proof-new) Implementation of trust substitution

### PR DESCRIPTION
**Pull request by ajreynol**
Trust substitution is a data structure that is used throughout preprocessing for proofs.

This adds its implementation.

### CODE REVIEW

File: src/theory/trust_substitutions.cpp [hunk 1]
```diff
@@ -20,41 +20,226 @@ namespace CVC4 {
...
+  // notice that this checks syntactic equality, not CDProof::isSame since
+  // tn.getGenerator() is not necessarily robust to symmetry.
+  if (eq == proven)
+  {
+    // no rewrite required, just use the generator
+    addSubstitution(x, t, tn.getGenerator());
+    Trace("trust-subs") << "...use generator directly" << std::endl;
     return;
   }
-  // NOTE: can try to transform tn.getProven() to (= x t) here
```
Comment (HanielB) [inline src/theory/trust_substitutions.cpp:51]: this comment was useful, restore it

File: src/theory/trust_substitutions.cpp [hunk 2]
```diff
@@ -20,41 +20,226 @@ namespace CVC4 {
...
+    // no proofs, use null generator
+    return TrustNode::mkTrustRewrite(n, ns, nullptr);
+  }
+  Node cs = getCurrentSubstitution();
+  Trace("trust-subs")
+      << "TrustSubstitutionMap::addSubstitution: current substitution is " << cs
+      << std::endl;
+  // Easy case: if n is in the domain of the substitution, maybe it is already
+  // a proof in the substitution proof generator. This is moreover required
+  // to avoid cyclic proofs below.
```
Comment (HanielB) [inline src/theory/trust_substitutions.cpp:166]: give an example

File: src/theory/trust_substitutions.cpp [hunk 3]
```diff
@@ -20,41 +20,226 @@ namespace CVC4 {
...
+  if (d_subsPg->hasStep(eq) || d_subsPg->hasGenerator(eq))
+  {
+    return TrustNode::mkTrustRewrite(n, ns, d_subsPg.get());
+  }
+  Assert(eq != cs);
+  std::vector<Node> pfChildren;
+  if (!cs.isConst())
+  {
+    // note we will get more proof reuse if we do not special case AND here.
+    if (false && cs.getKind() == kind::AND)
```
Comment (HanielB) [inline src/theory/trust_substitutions.cpp:177]: what is going on here?

**Review by HanielB — CHANGES_REQUESTED** [review]

File: src/theory/trust_substitutions.h [hunk 4]
```diff
@@ -52,6 +66,11 @@ class TrustSubstitutionMap
    * we simply add the substitution.
    */
   void addSubstitutionSolved(TNode x, TNode t, TrustNode tn);
+  /**
+   * Add substitutions from trust substitution map t. This adds all
+   * substitutions for t and carries over its information about proofs.
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/trust_substitutions.h:71]: Meant to say "from t".

File: src/theory/trust_substitutions.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/trust_substitutions.cpp:177]: I'll revert, I wanted to disable this block of code, but at the moment, its still necessary.

**Review by HanielB — APPROVED** [review]
LGTM, thanks for the changes.

**Pull request merged**